pkg/trace/pb/decoder.go

```diff
@@ -198,3 +236,225 @@ func parseInt32(dc *msgp.Reader) (int32, error) {
 		return 0, msgp.TypeError{Encoded: t, Method: msgp.IntType}
 	}
 }
+
+// DecodeMsgArray implements msgp.Decodable
+func (z *Traces) DecodeMsgArray(dc *msgp.Reader) (err error) {
+	if _, err := dc.ReadArrayHeader(); err != nil {
+		return err
+	}
+	// read dictionary
+	sz, err := dc.ReadArrayHeader()
+	if err != nil {
+		return err
+	}
+	dict := make([]string, sz)
+	for i := range dict {
+		str, err := parseString(dc)
+		if err != nil {
+			return err
+		}
+		dict[i] = str
+	}
+
+	// read traces
+	var xsz uint32
+	xsz, err = dc.ReadArrayHeader()
+	if err != nil {
+		return
+	}
+	if cap((*z)) >= int(xsz) {
+		(*z) = (*z)[:xsz]
+	} else {
+		(*z) = make(Traces, xsz)
+	}
+	for wht := range *z {
+		var xsz uint32
+		xsz, err = dc.ReadArrayHeader()
+		if err != nil {
+			return
+		}
+		if cap((*z)[wht]) >= int(xsz) {
+			(*z)[wht] = (*z)[wht][:xsz]
+		} else {
+			(*z)[wht] = make(Trace, xsz)
+		}
+		for hct := range (*z)[wht] {
+			if dc.IsNil() {
+				err = dc.ReadNil()
+				if err != nil {
+					return
+				}
+				(*z)[wht][hct] = nil
+			} else {
+				if (*z)[wht][hct] == nil {
+					(*z)[wht][hct] = new(Span)
+				}
+				err = (*z)[wht][hct].DecodeMsgArray(dc, dict)
+				if err != nil {
+					return
+				}
+			}
+		}
+	}
+	return
+}
+
+const spanPropertyCount = 12
+
+// DecodeMsgArray implements msgp.Decodable
+func (z *Span) DecodeMsgArray(dc *msgp.Reader, dict []string) (err error) {
+	var xsz uint32
+	xsz, err = dc.ReadArrayHeader()
+	if err != nil {
+		return
+	}
+	if xsz != spanPropertyCount {
+		return errors.New("encoded span needs exactly 12 elements in array")
+	}
+	for i := 0; i < spanPropertyCount; i++ {
```

Comment:
I'm confused by the need for a loop here.  Now that fields are in a known order shouldn't we be able to write this as a straight line piece of code.

In all likelihood, the compiler will eliminate the loop anyway, but I think it would just be cleaner without the loop.

Reply:
@richardstartin already made this comment and we’ve tried removing it. It serves no performance benefit. I prefer it like this with the numbered switch-cases. It’s easier to read.